From: Nathan Chancellor <natechancellor@gmail.com>
To: Jani Nikula <jani.nikula@linux.intel.com>,
    Joonas Lahtinen <joonas.lahtinen@linux.intel.com>,
    Rodrigo Vivi <rodrigo.vivi@intel.com>
Cc: intel-gfx@lists.freedesktop.org, dri-devel@lists.freedesktop.org,
    linux-kernel@vger.kernel.org, clang-built-linux@googlegroups.com,
    "Nathan Chancellor" <natechancellor@gmail.com>,
    "Michel Dänzer" <michel@daenzer.net>
Subject: [PATCH] drm/i915: Cast remain to unsigned long in eb_relocate_vma
Date: Thu, 13 Feb 2020 22:47:07 -0700
Message-ID: <20200214054706.33870-1-natechancellor@gmail.com> (raw)

A recent commit in clang added -Wtautological-compare to -Wall, which is
enabled for i915 after -Wtautological-compare is disabled for the rest
of the kernel so we see the following warning on x86_64:

 ../drivers/gpu/drm/i915/gem/i915_gem_execbuffer.c:1433:22: warning:
 result of comparison of constant 576460752303423487 with expression of
 type 'unsigned int' is always false
 [-Wtautological-constant-out-of-range-compare]
         if (unlikely(remain > N_RELOC(ULONG_MAX)))
            ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
 ../include/linux/compiler.h:78:42: note: expanded from macro 'unlikely'
 # define unlikely(x)    __builtin_expect(!!(x), 0)
                                            ^
 1 warning generated.

It is not wrong in the case where ULONG_MAX > UINT_MAX but it does not
account for the case where this file is built for 32-bit x86, where
ULONG_MAX == UINT_MAX and this check is still relevant.

Cast remain to unsigned long, which keeps the generated code the same
(verified with clang-11 on x86_64 and GCC 9.2.0 on x86 and x86_64) and
the warning is silenced so we can catch more potential issues in the
future.

Link: https://github.com/ClangBuiltLinux/linux/issues/778
Suggested-by: Michel Dänzer <michel@daenzer.net>
Signed-off-by: Nathan Chancellor <natechancellor@gmail.com>
---

Round 3 :)

Previous threads/patches:

https://lore.kernel.org/lkml/20191123195321.41305-1-natechancellor@gmail.com/
https://lore.kernel.org/lkml/20200211050808.29463-1-natechancellor@gmail.com/

 drivers/gpu/drm/i915/gem/i915_gem_execbuffer.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
diff -Naur linux-5.5/drivers/gpu/drm/i915/gem/i915_gem_execbuffer.c linux-5.5.tpg/drivers/gpu/drm/i915/gem/i915_gem_execbuffer.c
--- linux-5.5/drivers/gpu/drm/i915/gem/i915_gem_execbuffer.c	2020-01-27 00:23:03.000000000 +0000
+++ linux-5.5.tpg/drivers/gpu/drm/i915/gem/i915_gem_execbuffer.c	2020-03-15 13:48:28.187347096 +0000
@@ -1482,7 +1482,7 @@
 
 	urelocs = u64_to_user_ptr(entry->relocs_ptr);
 	remain = entry->relocation_count;
-	if (unlikely(remain > N_RELOC(ULONG_MAX)))
+	if (unlikely((unsigned long)remain > N_RELOC(ULONG_MAX)))
 		return -EINVAL;
 
 	/*
